integration-tests/chaos/congestion/simulator.go

```diff
@@ -0,0 +1,484 @@
+}
+
+func (s *Simulator) runTick(ctx context.Context) {
+	var wg sync.WaitGroup
+	ctx, cancel := context.WithTimeout(ctx, time.Second*time.Duration(s.RampUp.Duration+s.Plateau.Duration+s.CoolDown.Duration))
+	defer cancel()
+	wg.Add(1)
+	go s.runPhaseController(ctx, &wg)
+	s.startTransactionProducer(ctx, &wg)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		s.listenHeadsUntilCanceled(ctx, s.updateBaseFee)
+	}()
+	wg.Wait()
+}
+
+func (s *Simulator) updateBaseFee(_ context.Context, _ *types.Header) error {
+	err := s.anvilClient.AnvilSetNextBlockBaseFeePerGas(big.NewInt(0).SetUint64(s.expectedCurrentBaseFee.Load()))
+	if err != nil {
+		return fmt.Errorf("failed to set next block base fee per gas: %w", err)
+	}
+
+	return nil
+}
+
+func (s *Simulator) runSender(ctx context.Context, wg *sync.WaitGroup, key int, send <-chan struct{}) {
+	defer wg.Done()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-send:
+			_, err := s.emitter.EmitLogStringsFromKeyAsync(s.payload, key, func(o *bind.TransactOpts) {
+				o.GasLimit = s.Input.GasLimit
+				o.GasTipCap = big.NewInt(0).SetUint64(s.expectedCurrentTip.Load())
+				o.GasFeeCap = big.NewInt(0).Add(big.NewInt(0).SetUint64(s.expectedCurrentBaseFee.Load()), o.GasTipCap)
+			})
+			if err != nil {
+				s.lggr.Errorw("failed to emit log to fill block", "error", err)
+			}
+		}
+	}
+}
+
+func (s *Simulator) reportObservationOnNewHead(ctx context.Context, header *types.Header) error {
+	congestion := float64(header.GasUsed) / float64(header.GasLimit)
+	baseFeePercent := float64(header.BaseFee.Uint64()) / float64(s.Input.InitialBaseFeePerGas) * 100
+	block, err := s.client.Client.BlockByNumber(ctx, header.Number)
+	if err != nil {
+		return fmt.Errorf("failed to fetch block by number: %w", err)
+	}
+	tipPercent := float64(s.avgTip(block).Uint64()) / float64(s.InitialTipPerGas) * 100
+
+	state := chainState{
+		PhaseName:           s.phase.Load().(phaseType),
+		Congestion:          congestion,
+		TipDeltaPercent:     tipPercent - 100,
+		BaseFeeDeltaPercent: baseFeePercent - 100,
+	}
+	if s.observationsHandler != nil {
+		s.observationsHandler(state)
+	}
+
+	s.lggr.Debugw("New chain state report", "chainState", state)
+	return nil
+}
+
+func (s *Simulator) avgTip(block *types.Block) *big.Int {
+	if block.Transactions().Len() == 0 {
+		return big.NewInt(0)
+	}
+	total := big.NewInt(0)
+	for _, tx := range block.Transactions() {
+		total = total.Add(tx.EffectiveGasTipValue(block.BaseFee()), total)
+	}
+
+	return big.NewInt(0).Div(total, big.NewInt(int64(block.Transactions().Len())))
+}
+
+func (s *Simulator) listenHeadsUntilCanceled(ctx context.Context, onNewBlock func(ctx context.Context, head *types.Header) error) {
+	for {
+		err := s.listenHeads(ctx, onNewBlock)
+		if err != nil {
+			if ctx.Err() != nil {
+				return
+			}
+
+			s.lggr.Errorw("failed to run onNewBlock", "error", err)
+		}
+		time.Sleep(backoffPeriodOnError)
```

Comment:
shouldn't that be inside the `if err != nil` brackets? Or is that parameter `backoffPeriodOnError` incorrectly named?